### PR TITLE
Create project proposal for CoreDNS

### PR DESCRIPTION
**Pull request by caniszczyk**
Initial draft of the project proposal for CoreDNS

https://github.com/miekg/coredns

### CODE REVIEW

File: proposals/coredns.adoc [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+* Felix Cantournet github: fcantournet
+* github: leelynne
+* Matt Layher github: mdlayher
+* Vasily Vailyev github: pixelbender
+
+*Infrastructure requirements (CI / CNCF Cluster)*: N/A
+
+*Issue tracker*: https://github.com/miekg/coredns
+
+*Website*:	https://miek.nl/tags/coredns/
```
Comment (miekg) [inline proposals/coredns.adoc]: https://coredns.io

File: proposals/coredns.adoc [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+
+*Release methodology and mechanics*:	github master
+
+*Social media accounts*:	Twitter: @coredns
```
Comment (miekg) [inline proposals/coredns.adoc]: @corednsio

**Comment by caniszczyk** [comment]
Thanks @miekg, updated

File: proposals/coredns.adoc [hunk 3]
```diff
@@ -0,0 +1,82 @@
+== CoreDNS Proposal
+
+*Name of project:* CoreDNS
+
+*Description*
+
+CoreDNS is a fast, flexible DNS server that is designed to be performant with a flexible implementation. This flexibility allows CoreDNS to be easily extended to support various data sources and to implement rich DNS service behaviors. For example, response caching, query rewrite, load-balancing, zone transfer and signing.
```
Comment (jonboulle) [inline proposals/coredns.adoc]: behaviors: for example,

File: proposals/coredns.adoc [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+CoreDNS is the successor of SkyDNS.
```
Comment (jonboulle) [inline proposals/coredns.adoc]: A touch more background/context about SkyDNS might be nice.

File: proposals/coredns.adoc [hunk 5]
```diff
@@ -0,0 +1,82 @@
...
+* Matt Layher github: mdlayher
+* Vasily Vailyev github: pixelbender
+
+*Infrastructure requirements (CI / CNCF Cluster)*: N/A
+
+*Issue tracker*: https://github.com/miekg/coredns
+
+*Website*: https://coredns.io
+
+*Release methodology and mechanics*: github master
```
Comment (jonboulle) [inline proposals/coredns.adoc]: Flesh this out a bit. It looks like there are no releases so far - but perhaps talk about why, and what process will be introduced (semver? prebuilt binaries?)

File: proposals/coredns.adoc [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+* Felix Cantournet github: fcantournet
+* github: leelynne
+* Matt Layher github: mdlayher
+* Vasily Vailyev github: pixelbender
+
+*Infrastructure requirements (CI / CNCF Cluster)*: N/A
+
+*Issue tracker*: https://github.com/miekg/coredns
+
+*Website*: https://coredns.io
```
Comment (jonboulle) [inline proposals/coredns.adoc]: nice inspirational looking site :-)

File: proposals/coredns.adoc [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+*Release methodology and mechanics*: github master
+
+*Social media accounts*: Twitter: @corednsio
+
+*Existing sponsorship*: Infoblox contributing developer time to implement CoreDNS->Kubernetes integration component.
+
+*External Dependencies*
```
Comment (jonboulle) [inline proposals/coredns.adoc]: Could you talk a bit about the relationship with Caddy? from our discussion on the call the other day it sounded like Caddy will become a dependency once CoreDNS is ported over (which looks like it's happening [soon](https://github.com/miekg/coredns/pull/220)?)

File: proposals/coredns.adoc [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+* Go package: golang/x/crypto (BSD https://github.com/golang/crypto/blob/master/LICENSE)
+* Go package: golang/x/net (BSD https://github.com/golang/net/blob/master/LICENSE)
+* Go package: golang/x/sys (BSD https://github.com/golang/sys/blob/master/LICENSE) 
+* Go package: natefinch/lumberjack.v2 (MIT https://github.com/natefinch/lumberjack/blob/v2.0/LICENSE) 
+* Go package: square/go-jose.v1 (ASLv2 https://github.com/square/go-jose/blob/master/LICENSE)
+* Kubernetes (for CoreDNS -> kubernetes integration) (ASLv2 https://github.com/kubernetes/kubernetes/blob/master/LICENSE) 
+
+*Statement on alignment with CNCF mission*:
+ 
+CoreDNS is a focused lightweight DNS server. A microservice philosophy guides the internal design of CoreDNS. Individual DNS functions are provided by discrete, composable plugins that are enabled via runtime configuration.
```
Comment (jonboulle) [inline proposals/coredns.adoc]: `..of CoreDNS: individual..`

File: proposals/coredns.adoc [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+CoreDNS can be thought of as a DNS protocol head that can be configured to front various backend data sources. A flexible DNS server is a necessary component to provide “Naming and Discovery” services to containers running in the CNCF distributed system services environment.
```
Comment (jonboulle) [inline proposals/coredns.adoc]: perhaps "frontend" instead of "head"

Comment (jonboulle) [inline proposals/coredns.adoc]: A flexible DNS server is a necessary component in providing “Naming and Discovery” services to containers running in a cloud-native distributed environment.

**Comment by jonboulle** [comment]
Mostly LGTM; the big outstanding question to me is the caddy stuff, I'd like to be a bit clearer here about the future plans/implications there.

**Comment by jonboulle** [comment]
Also, could you add a note about existing community size/adoption/etc?

File: proposals/coredns.adoc (hunk 4 above)
Comment (miekg) [inline proposals/coredns.adoc]: PR #10 send to resolve this and the dep. on Caddy.

**Comment by miekg** [comment]
Send two PRs today, I think they should address all of Jon's comments.

Also Jon: Hi! And thanks for sponsering @jonboulle :-) Feel free to say if this isn't sufficient yet.

@mrichmon anything you want to add?

**Comment by mrichmon** [comment]
Opened a PR to add miekg to the initial committers list.

Otherwise, this looks like a cut and paste from the Google Document that I wrote. So I'm good with this. Caddy is essentially just another framework dependency.

**Comment by jonboulle** [comment]
I put up #14 but it's mostly nits/grammar cleanup. Overall this LGTM!

**Comment by jonboulle** [comment]
:+1:

**Comment by caniszczyk** [comment]
If this is ready to vote on @jonboulle I'd recommend doing a +1 here, would love to see if we can use PullApprove to automatically count the minimum 6 +1 votes before we merge in

**Comment by johnbelamaric** [comment]
Please note that CoreDNS is being proposed as an *inception* project per https://github.com/cncf/toc/blob/master/process/graduation_criteria.adoc.

**Comment by jonboulle** [comment]
@johnbelamaric mind adding that to the doc? c.f. https://github.com/cncf/toc/pull/27

**Comment by johnbelamaric** [comment]
no problem

**Comment by johnbelamaric** [comment]
In PR #28

File: proposals/coredns.adoc [hunk 10]
```diff
@@ -0,0 +1,103 @@
...
+A more formal release process is on its way, and may introduce semantic versioning, but a final decision has not yet been made. Precompiled binaries are
+distributed via GitHub and Caddy's download website (https://caddyserver.com/download), where "DNS" will be a Server Type option.
+
+*Social media accounts*: Twitter: @corednsio
+
+*Existing sponsorship*: Infoblox contributing developer time to implement CoreDNS->Kubernetes integration component and other enhancements.
+
+*Existing community*: The community is small, but growing. On GitHub, we have 50+ forks, 550+ stars, and 24 contributors (5 with merge privileges).
+Current number of Twitter followers is 180+. By aligning ourselves with the Caddy community, we hope to leverage Caddy's
+popularity for CoreDNS. By positioning CoreDNS as a better SkyDNS, we hope to entice existing users of SkyDNS to migrate to and embrace CoreDNS.
```
Comment (jonboulle) [inline proposals/coredns.adoc]: Might be worth calling out if there is a migration path/guidance for existing SkyDNS users

File: proposals/coredns.adoc [hunk 11]
```diff
@@ -0,0 +1,103 @@
...
+*External Dependencies*
+
+CoreDNS depends on Caddy (https://caddyserver.com/) as a standard Go dependency. Caddy is a framework that CoreDNS uses in two ways:
+
+1. Much of the CoreDNS code plugs into the framework to add DNS behavior.
+2. CoreDNS provides a wrapper around the framework to provide a DNS-tuned command-line interface.
+
+Go dependencies:
+
+* Go package: mholt/caddy (ASLV2 https://github.com/mholt/caddy/blob/master/LICENSE.txt)
```
Comment (jonboulle) [inline proposals/coredns.adoc]: sort this list? appears somewhat random

File: proposals/coredns.adoc [hunk 12]
```diff
@@ -0,0 +1,103 @@
...
+
+*Statement on alignment with CNCF mission*:
+
+CoreDNS is a focused, lightweight DNS server. A microservice philosophy guides the internal design of CoreDNS. Individual DNS functions are provided by discrete, composable plugins that are enabled via runtime configuration.
+CoreDNS can be thought of as a DNS protocol head that can be configured to front various backend data sources. A flexible DNS server is a necessary component to provide “Naming and Discovery” services to containers running in the CNCF distributed system services environment.
+
+*Comparison with KubeDNS*:
+
+The incumbent DNS service for Kubernetes, “kubedns”, consists of three components:
+* kube-dns which uses SkyDNS as a library provides the DNS service based on the Kubernetes API
```
Comment (jonboulle) [inline proposals/coredns.adoc]: kube-dns, which uses SkyDNS as a library, provides ...

File: proposals/coredns.adoc [hunk 13]
```diff
@@ -16,6 +16,8 @@ used in cloud deployments, but lacks the flexibility of CoreDNS.
 
 *License*: Apache License v2.0
 
+*Maturity Level*: Inception
```
Comment (johnbelamaric) [inline proposals/coredns.adoc]: *Preferred Maturity Level* is above, should it be removed?

**Comment by johnbelamaric** [comment]
👍

**Comment by jonboulle** [comment]
:+1:

**Pull request merged**